quickwit-search/src/root.rs

```diff
@@ -136,39 +142,60 @@ fn analyze_errors(search_result: &SearchResultsByAddr) -> Option<ErrorRetries> {
         .filter(|(_addr, result)| {
             result
                 .as_ref()
-                .map(|ok_res| ok_res.failed_requests.len() as u64 != ok_res.num_attempted_splits)
+                .map(|ok_res| ok_res.failed_splits.len() as u64 != ok_res.num_attempted_splits)
                 .unwrap_or(false)
         })
         .map(|(addr, _)| *addr)
         .collect();
+    debug!("{:?}", &complete_failure_nodes);
+    debug!("{:?}", &partial_or_no_failure_nodes);
 
     // Here we collect the failed requests on the node. It does not yet include failed requests
     // against that node
-    let failed_split_ids = search_result
+    let mut retry_split_ids = search_result
         .values()
         .filter_map(|result| result.as_ref().ok())
         .flat_map(|res| {
             let results = res
-                .failed_requests
+                .failed_splits
                 .iter()
-                .map(move |failed_req| failed_req.split_id.to_string())
+                .filter(move |failed_splits| failed_splits.retryable_error) // only retry splits marked as retryable
+                .map(move |failed_splits| failed_splits.split_id.to_string())
                 .collect_vec();
             results.into_iter()
         })
         .collect_vec();
 
-    let retry_nodes = if partial_or_no_failure_nodes.is_empty() {
-        complete_failure_nodes
-    } else {
-        partial_or_no_failure_nodes
-    };
+    let failed_splits = search_result
+        .values()
+        .filter_map(|result| result.as_ref().err())
+        .flat_map(|err| err.split_ids.iter().cloned());
+    retry_split_ids.extend(failed_splits);
+
+    let nodes_to_avoid = complete_failure_nodes;
 
     Some(ErrorRetries {
-        retry_split_ids: failed_split_ids,
-        nodes_to_use: retry_nodes,
+        retry_split_ids,
+        nodes_to_avoid: nodes_to_avoid.into_iter().collect(),
     })
 }
 
+fn job_for_split(
+    split_ids: &HashSet<&String>,
+    split_metadata_map: &HashMap<String, SplitMetadata>,
+) -> Vec<Job> {
+    // Create a job for fetching docs and assign the splits that the node is responsible for based on the job.
+    let leaf_search_jobs: Vec<Job> = split_metadata_map
+        .iter()
+        .filter(|(split_id, _)| split_ids.contains(split_id))
+        .map(|(split_id, split_metadata)| Job {
+            split: split_id.clone(),
+            cost: compute_split_cost(split_metadata),
+        })
+        .collect();
+    leaf_search_jobs
+}
+
 /// Perform a distributed search.
 /// It sends a search request over gRPC to multiple leaf nodes and merges the search results.
 pub async fn root_search(
```

Comment:
I think you wrote somewhere a great explanation on your retry logic.

Can you copy paste it in the doc of root_search and shorten it if necessary.

If I understand correctly, 


After a first round of leaf_search requests, we identify the list of failed but retryable splits,
and retry them. The leaf nodes which did not return any result are suspected to be unhealthy and are excluded
from this retry round.